---
title: "05-cell_marker_genes"
author: "Melanie Smith"
date: "2026-01-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Clean-up environment

```{r}
rm(list = ls(all.names = TRUE)) # will clear all objects includuing hidden objects
gc() # free up memory and report the memory usage
options(max.print = .Machine$integer.max,
        scipen = 999,
        stringsAsFactors = FALSE,
        dplyr.summarise.inform = FALSE) # Print everything without truncation, avoid scientific notation, keep strings as characters, silence dplyr’s “grouped output” informational messages
set.seed(42) # set seed for reproducibility
```

# Load Libraries

```{r}
# Core tidyverse
library(tidyverse)
library(tibble)
library(dplyr)
library(tidyr)

# Single Cell
library(Seurat) # Seurat (relies on dplyr/tidyr, so load after tidyverse)

library(SingleCellExperiment)
library(biomaRt)
library(scuttle)

# additional plotting functions
library(patchwork)
library(pheatmap)
```

# User defined functions

```{r}

```

# Set project directories and file paths

```{r}
project_name <- "20251104_sophieWiszniak_ncc_pa"

base_repo_dir <- "/home/melanie-smith/workDir/sophieWiszniak"
project_repo_dir <- file.path(base_repo_dir, project_name)
data_base_dir <- file.path(project_repo_dir, "rawData")
out_dir <- file.path(project_repo_dir, "outDir")
downloads <- file.path(project_repo_dir, "downloads")

new_out_dir <- file.path(out_dir, "05-secondary_QC")

```

# Load required objects
- This object has been filtered for doublets, >10% MT genes and low quality clusters have been removed

```{r}
re_embed_no_low.qual <- readRDS(file = file.path(new_out_dir, "filtered_mt_doublet_nFeature_top.low_quality.rds"))
```

## Plot cell-type marker genes
nb: The marker genes used here are from [Single-cell RNA-Seq of the developing cardiac outflow tract reveals convergent development of the vascular smooth muscle cells](https://www.sciencedirect.com/science/article/pii/S2211124719308721?via%3Dihub#fig5)

- Macrophage

```{r}
macrophage_markers <- FeaturePlot(
  object = re_embed_no_low.qual,
  features = c("Fcgr1", "Adgre1"),
  ncol = 2,                      # side-by-side panels
  order = TRUE                   # high-expression cells on top
  ) &
  scale_colour_gradientn(
    colours = rev(RColorBrewer::brewer.pal(11, "Spectral")),
    name = "Expression"
  ) &
  plot_annotation(
    title = "Macrophage markers",
    caption = "Filtered: <10% MT, singlets only, nFeatures >0.5%, no low quality"
  ) &
  theme_bw(base_size = 14) &
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    plot.subtitle = element_text(hjust = 0.5)
  )
ggsave(macrophage_markers, 
       filename = file.path(new_out_dir, "05-umap.macrophage_genes.png"),
       width = 16,
       height = 8)
```

- Epicardial lineage

```{r}
epicardial_markers <- FeaturePlot(
  object = re_embed_no_low.qual,
  features = c("Upk3b", "Upk1b"),
  ncol = 2,                      # side-by-side panels
  order = TRUE                   # high-expression cells on top
  ) &
  scale_colour_gradientn(
    colours = rev(RColorBrewer::brewer.pal(11, "Spectral")),
    name = "Expression"
  ) &
  plot_annotation(
    title = "Epicardial lineage markers",
    caption = "Filtered: <10% MT, singlets only, nFeatures >0.5%, no low quality"
  ) &
  theme_bw(base_size = 14) &
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    plot.subtitle = element_text(hjust = 0.5)
  )
ggsave(epicardial_markers, 
       filename = file.path(new_out_dir, "05-umap.epicardial_genes.png"),
       width = 16,
       height = 8)
```

- Mesenchymal lineage

```{r}
mesenchymal_markers <- FeaturePlot(
  object = re_embed_no_low.qual,
  features = c("Postn", "Cthrc1"),
  ncol = 2,                      # side-by-side panels
  order = TRUE                   # high-expression cells on top
  ) &
  scale_colour_gradientn(
    colours = rev(RColorBrewer::brewer.pal(11, "Spectral")),
    name = "Expression"
  ) &
  plot_annotation(
    title = "Mesenchymal lineage markers",
    caption = "Filtered: <10% MT, singlets only, nFeatures >0.5%, no low quality"
  ) &
  theme_bw(base_size = 14) &
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    plot.subtitle = element_text(hjust = 0.5)
  )
ggsave(mesenchymal_markers, 
       filename = file.path(new_out_dir, "umap.mesenchymal_genes.png"),
       width = 16,
       height = 8)
```

- Myocardial lineage

```{r}
myocardial_markers <- FeaturePlot(
  object = re_embed_no_low.qual,
  features = c("Myh7", "Myl4"),
  ncol = 2,                      # side-by-side panels
  order = TRUE                   # high-expression cells on top
  ) &
  scale_colour_gradientn(
    colours = rev(RColorBrewer::brewer.pal(11, "Spectral")),
    name = "Expression"
  ) &
  plot_annotation(
    title = "Myocardial lineage markers",
    caption = "Filtered: <10% MT, singlets only, nFeatures >0.5%, no low quality"
  ) &
  theme_bw(base_size = 14) &
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    plot.subtitle = element_text(hjust = 0.5)
  )
ggsave(myocardial_markers, 
       filename = file.path(new_out_dir, "05-umap.myocardial_genes.png"),
       width = 16,
       height = 8)
```

- Endocardial lineage

```{r}
endocardial_markers <- FeaturePlot(
  object = re_embed_no_low.qual,
  features = c("Ecscr", "Cdh5"),
  ncol = 2,                      # side-by-side panels
  order = TRUE                   # high-expression cells on top
  ) &
  scale_colour_gradientn(
    colours = rev(RColorBrewer::brewer.pal(11, "Spectral")),
    name = "Expression"
  ) &
  plot_annotation(
    title = "Endocardial lineage markers",
    caption = "Filtered: <10% MT, singlets only, nFeatures >0.5%, no low quality"
  ) &
  theme_bw(base_size = 14) &
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    plot.subtitle = element_text(hjust = 0.5)
  )
ggsave(endocardial_markers, 
       filename = file.path(new_out_dir, "05-umap.endocardial_genes.png"),
       width = 16,
       height = 8)
```

- VSMC lineage

```{r}
vsmc_markers <- FeaturePlot(
  object = re_embed_no_low.qual,
  features = c("Cxcl12", "Rgs5"),
  ncol = 2,                      # side-by-side panels
  order = TRUE                   # high-expression cells on top
  ) &
  scale_colour_gradientn(
    colours = rev(RColorBrewer::brewer.pal(11, "Spectral")),
    name = "Expression"
  ) &
  plot_annotation(
    title = "VSMC lineage markers",
    caption = "Filtered: <10% MT, singlets only, nFeatures >0.5%, no low quality"
  ) &
  theme_bw(base_size = 14) &
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    plot.subtitle = element_text(hjust = 0.5)
  )
ggsave(vsmc_markers, 
       filename = file.path(new_out_dir, "05-umap.vsmc_genes.png"),
       width = 16,
       height = 8)
```

- RBC lineage (taken from Nick's hb genes)

```{r}
rbc_markers <- FeaturePlot(
  object = re_embed_no_low.qual,
  features = c("Hba-a1", "Hba-a2"),
  ncol = 2,                      # side-by-side panels
  order = TRUE                   # high-expression cells on top
  ) &
  scale_colour_gradientn(
    colours = rev(RColorBrewer::brewer.pal(11, "Spectral")),
    name = "Expression"
  ) &
  plot_annotation(
    title = "RBC lineage markers",
    caption = "Filtered: <10% MT, singlets only, nFeatures >0.5%, no low quality"
  ) &
  theme_bw(base_size = 14) &
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    plot.subtitle = element_text(hjust = 0.5)
  )
ggsave(rbc_markers, 
       filename = file.path(new_out_dir, "05-umap.rbc_genes.png"),
       width = 16,
       height = 8)
```

- Platelet lineage (taken from Nick's hb genes)

```{r}
platelet_markers <- FeaturePlot(
  object = re_embed_no_low.qual,
  features = c("Pf4", "Ppbp"),
  ncol = 2,                      # side-by-side panels
  order = TRUE                   # high-expression cells on top
  ) &
  scale_colour_gradientn(
    colours = rev(RColorBrewer::brewer.pal(11, "Spectral")),
    name = "Expression"
  ) &
  plot_annotation(
    title = "Platelet lineage markers",
    caption = "Filtered: <10% MT, singlets only, nFeatures >0.5%, no low quality"
  ) &
  theme_bw(base_size = 14) &
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    plot.subtitle = element_text(hjust = 0.5)
  )
ggsave(platelet_markers, 
       filename = file.path(new_out_dir, "05-umap.platelet_genes.png"),
       width = 16,
       height = 8)
```

## Dot plots of lineage markers
- First "layer" of markers from the paper (ie first level of boxes)

```{r}
# lets start with a list of broad lineage marker genes
macrophage_genes <- c("Coro1a", "C3ar1", "Rac2", "C1qb", "Tyrobp")

epicardial_genes <- c("Wt1", "Upk3b", "Apela", "Upk1b", "Aldh1a2")
ep_c10 <- c("Myl7", "Tmem255a")
ep_c11 <- c("Eln", "Dlk1", "Klk14")

endocardial_genes <- c("Ecscr", "Cdh5", "Tie1", "Rasip1", "Plxnd1")
ed_c5 <- c("Crip1", "Hapln1", "Igfbp7")
ed_c6 <- c("Rbp1", "Crip1", "Fbln5", "Gja4")
ed_c12 <- c("Emcn", "klhl4", "Gatm", "Akap12")

myocardial_genes <- c("Chchd10", "Sh3bgr", "Ttn", "Pln", "Myh7")
mc_c2 <- c("Myl2", "Atp1b1", "Cox6a2")
mc_c9 <- c("Bmp4", "Malat1", "Gpc3")

mesenchymal_genes <- c("Postn", "Vcan", "Dbi", "Cthrc1", "Papss2")
ms_c0 <- c("Itga4", "Pla2g16", "Myl2")
ms_c1 <- c("Htra1", "Eln")
ms_c7 <- c("Actb", "Cbx3", "Kdelr2", "Slc39a1")
ms_c8 <- c("Penk", "Cxcl12", "Fbln5", "Wnt5a")

vsmc_genes <- c("Cxcl12", "Iigp1", "Eln", "Fbln5", "Rgs5")
vsmc_c3 <- c("Sfrp2", "Osr1", "Bgn", "Cd34")
vsmc_c6 <- c("Rbp1", "Fbln5", "Gja4")
vsmc_c12 <- c("Emcn", "Klhl4", "Gatm", "Akap12")
```

## Macrophage dotplot

```{r}
# Generate DotPlot for original_clusters_mt_dbl markers
Idents(re_embed_no_rbc) <- "original_clusters_mt_dbl"
dotPlot_original_clusters_macrophage <- DotPlot(re_embed_no_rbc,
        features = macrophage_genes,
        dot.scale = 6,
        cols = c("blue", "red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(
    title = "Macrophage Genes",
    subtitle = "Filtered: <10% MT, singlets only, RBC removed",
    caption = "Mib1 KO vs control"
    )
   
ggsave(dotPlot_original_clusters_macrophage,
       filename = file.path(out_dir, "05-secondary_QC", "05-dotPlot_original_clusters_macrophage.png"),
       width = 16,
       height = 6)
```

## Epicardial dotplot

```{r}
dotPlot_original_clusters_epicardial <- DotPlot(re_embed_no_rbc,
        features = epicardial_genes,
        dot.scale = 6,
        cols = c("blue", "red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(
    title = "Epicardial lineage Genes",
    subtitle = "Filtered: <10% MT, singlets only, RBC removed",
    caption = "Mib1 KO vs control"
    )
   
ggsave(dotPlot_original_clusters_epicardial,
       filename = file.path(out_dir, "05-secondary_QC", "05-dotPlot_original_clusters_epicardial.png"),
       width = 16,
       height = 6)

# all the epicardial genes
epicardial_plus <- c(epicardial_genes, ep_c10, ep_c11)
dotPlot_epicardial_plus <- DotPlot(re_embed_no_rbc,
        features = epicardial_plus,
        dot.scale = 6,
        cols = c("blue", "red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(
    title = "Epicardial (plus) lineage Genes",
    subtitle = "Filtered: <10% MT, singlets only, RBC removed",
    caption = "Mib1 KO vs control"
    )
   
ggsave(dotPlot_epicardial_plus,
       filename = file.path(out_dir, "05-secondary_QC", "05-dotPlot_epicardial_plus.png"),
       width = 16,
       height = 6)
```

## Endocardial dotplot

```{r}
dotPlot_original_clusters_endocardial <- DotPlot(re_embed_no_rbc,
        features = endocardial_genes,
        dot.scale = 6,
        cols = c("blue", "red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(
    title = "Endocardial lineage Genes",
    subtitle = "Filtered: <10% MT, singlets only, RBC removed",
    caption = "Mib1 KO vs control"
    )
   
ggsave(dotPlot_original_clusters_endocardial,
       filename = file.path(out_dir, "05-secondary_QC", "05-dotPlot_original_clusters_endocardial.png"),
       width = 16,
       height = 6)

# all the endocardial genes
endocardial_plus <- c(endocardial_genes, ed_c5, ed_c6, ed_c12)
dotPlot_endocardial_plus <- DotPlot(re_embed_no_rbc,
        features = unique(endocardial_plus),
        dot.scale = 6,
        cols = c("blue", "red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(
    title = "Endocardial (plus) lineage Genes",
    subtitle = "Filtered: <10% MT, singlets only, RBC removed",
    caption = "Mib1 KO vs control"
    )
   
ggsave(dotPlot_endocardial_plus,
       filename = file.path(out_dir, "05-secondary_QC", "05-dotPlot_endocardial_plus.png"),
       width = 16,
       height = 6)
```

## Myocardial dotplots

```{r}
dotPlot_original_clusters_myocardial <- DotPlot(re_embed_no_rbc,
        features = myocardial_genes,
        dot.scale = 6,
        cols = c("blue", "red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(
    title = "Myocardial lineage Genes",
    subtitle = "Filtered: <10% MT, singlets only, RBC removed",
    caption = "Mib1 KO vs control"
    )
   
ggsave(dotPlot_original_clusters_myocardial,
       filename = file.path(out_dir, "05-secondary_QC", "05-dotPlot_original_clusters_myocardial.png"),
       width = 16,
       height = 6)

# all the myocardial genes
myocardial_plus <- c(myocardial_genes, mc_c2, mc_c9)
dotPlot_myocardial_plus <- DotPlot(re_embed_no_rbc,
        features = unique(myocardial_plus),
        dot.scale = 6,
        cols = c("blue", "red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(
    title = "Myocardial (plus) lineage Genes",
    subtitle = "Filtered: <10% MT, singlets only, RBC removed",
    caption = "Mib1 KO vs control"
    )
   
ggsave(dotPlot_myocardial_plus,
       filename = file.path(out_dir, "05-secondary_QC", "05-dotPlot_myocardial_plus.png"),
       width = 16,
       height = 7)
```

## Mesenchymal dotplots

```{r}
dotPlot_original_clusters_mesenchymal <- DotPlot(re_embed_no_rbc,
        features = mesenchymal_genes,
        dot.scale = 6,
        cols = c("blue", "red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(
    title = "Mesenchymal lineage Genes",
    subtitle = "Filtered: <10% MT, singlets only, RBC removed",
    caption = "Mib1 KO vs control"
    )
   
ggsave(dotPlot_original_clusters_mesenchymal,
       filename = file.path(out_dir, "05-secondary_QC", "05-dotPlot_original_clusters_mesenchymal.png"),
       width = 16,
       height = 6)

# all the mesenchymal genes
mesenchymal_plus <- c(mesenchymal_genes, ms_c0, ms_c1, ms_c7, ms_c8)
dotPlot_mesenchymal_plus <- DotPlot(re_embed_no_rbc,
        features = unique(mesenchymal_plus),
        dot.scale = 6,
        cols = c("blue", "red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(
    title = "Mesenchymal (plus) lineage Genes",
    subtitle = "Filtered: <10% MT, singlets only, RBC removed",
    caption = "Mib1 KO vs control"
    )
   
ggsave(dotPlot_mesenchymal_plus,
       filename = file.path(out_dir, "05-secondary_QC", "05-dotPlot_mesenchymal_plus.png"),
       width = 16,
       height = 7)
```

## VSMC dotplots

```{r}
dotPlot_original_clusters_vsmc <- DotPlot(re_embed_no_rbc,
        features = vsmc_genes,
        dot.scale = 6,
        cols = c("blue", "red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(
    title = "VSMC lineage Genes",
    subtitle = "Filtered: <10% MT, singlets only, RBC removed",
    caption = "Mib1 KO vs control"
    )
   
ggsave(dotPlot_original_clusters_vsmc,
       filename = file.path(out_dir, "05-secondary_QC", "05-dotPlot_original_clusters_vsmc.png"),
       width = 16,
       height = 6)

# all the VSMC genes
vsmc_plus <- c(vsmc_genes, vsmc_c3, vsmc_c6, vsmc_c12)
dotPlot_vsmc_plus <- DotPlot(re_embed_no_rbc,
        features = unique(vsmc_plus),
        dot.scale = 6,
        cols = c("blue", "red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(
    title = "VSMC (plus) lineage Genes",
    subtitle = "Filtered: <10% MT, singlets only, RBC removed",
    caption = "Mib1 KO vs control"
    )
   
ggsave(dotPlot_vsmc_plus,
       filename = file.path(out_dir, "05-secondary_QC", "05-dotPlot_vsmc_plus.png"),
       width = 16,
       height = 7)

```

## All markers
- level one boxes only

```{r}
dotPlot_markers <- DotPlot(re_embed_no_rbc,
        features = c(mesenchymal_genes,
                     vsmc_genes,
                     epicardial_genes,
                     endocardial_genes,
                     macrophage_genes,
                     myocardial_genes
                     ),
        dot.scale = 6,
        cols = c("blue", "red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(
    title = "Level 1 lineage Genes",
    subtitle = "Filtered: <10% MT, singlets only, RBC removed",
    caption = "Mib1 KO vs control"
    )
print(dotPlot_markers)   
ggsave(dotPlot_markers,
       filename = file.path(out_dir, "05-secondary_QC", "05-dotPlot_markers.png"),
       width = 16,
       height = 8)

# all the marker genes
all_markers <- c(mesenchymal_plus,
                 vsmc_plus,
                 epicardial_plus,
                 endocardial_plus,
                 macrophage_genes,
                 myocardial_plus
                 )
dotPlot_all_markers <- DotPlot(re_embed_no_rbc,
        features = unique(all_markers),
        dot.scale = 6,
        cols = c("blue", "red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(
    title = "Combined (aka all level) cell-type marker genes",
    subtitle = "Filtered: <10% MT, singlets only, RBC removed",
    caption = "Mib1 KO vs control"
    )
   
ggsave(dotPlot_all_markers,
       filename = file.path(out_dir, "05-secondary_QC", "05-dotPlot_all_markers.png"),
       width = 20,
       height = 12)

```